File: go-client/client_lib/retry_executor.go
package clientlib

import (
	"fmt"
	"time"
)

type RetryableOperation func() (int, error)

type RetryExecutor struct {
	MaxAttempts   int
	InitialDelay  time.Duration
	MaxDelay      time.Duration
}

func NewRetryExecutor(maxAttempts int, initialDelay, maxDelay time.Duration) *RetryExecutor {
	return &RetryExecutor{
		MaxAttempts:  maxAttempts,
		InitialDelay: initialDelay,
		MaxDelay:     maxDelay,
	}
}

func NewDefaultRetryExecutor() *RetryExecutor {
	return NewRetryExecutor(3, 3*time.Second, 9*time.Second)
}

func (r *RetryExecutor) ExecuteWithRetry(operation RetryableOperation) (int, error) {
	var lastErr error
	attempt := 0
	delay := r.InitialDelay
	for attempt < r.MaxAttempts { 
		attempt++
		fmt.Println()
		fmt.Printf("Tentativa %d de %d\n", attempt, r.MaxAttempts)

		result, err := operation()
		if err == nil {
			return result, nil 
		}

		lastErr = err
		fmt.Println()
		fmt.Printf("Tentativa %d falhou: %v\n", attempt, err)

		if attempt == r.MaxAttempts {
			break 
		}

		fmt.Println()
		fmt.Printf("Aguardando %v antes da próxima tentativa...\n", delay)

		time.Sleep(delay)
		delay = r.calculaDelay(attempt)
	}

	return -1, fmt.Errorf("Todas as %d tentativas falharam. Último erro: %v", r.MaxAttempts, lastErr)
}

func (r *RetryExecutor) calculaDelay(attempt int) time.Duration {
	delay := time.Duration(float64(r.InitialDelay) * 2) // dobra o tempo de espera
	if delay > r.MaxDelay {
		return r.MaxDelay
	}
	return delay
}

File: java/src/main/java/com/mycompany/ine5418/client_lib/RetryExecutor.java
package com.mycompany.ine5418.client_lib;

import java.util.concurrent.TimeUnit;

public class RetryExecutor {
    private final int maxAttempts;
    private final long initialDelayMs;
    private final long maxDelayMs;

    public RetryExecutor(int maxAttempts, long initialDelayMs, long maxDelayMs) {
        this.maxAttempts = maxAttempts;
        this.initialDelayMs = initialDelayMs;
        this.maxDelayMs = maxDelayMs;
    }

    // Construtor com valores padrão
    public RetryExecutor() {
        this(3, 1500, 3000); // 3 tentativas, começando com 100ms, máximo 1s
    }

    public <T> T executeWithRetry(RetryableOperation<T> operation) {
        int attempt = 0;
        Exception lastException = null;

        while (attempt < maxAttempts) {
            try {
                attempt++;
                System.out.println("Tentativa " + attempt + " de " + maxAttempts);
                return operation.execute();

            } catch (Exception e) {
                lastException = e;
                System.out.println("Tentativa " + attempt + " falhou: " + e.getMessage());

                if (attempt == maxAttempts) {
                    break; // Última tentativa falhou
                }

                // Calcula delay com backoff exponencial
                long delay = calculateBackoffDelay(attempt);
                System.out.println("Aguardando " + delay + "ms antes da próxima tentativa...");

                try {
                    Thread.sleep(delay);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    throw new RuntimeException("Operação interrompida", ie);
                }
            }
        }

        throw new RuntimeException("Todas as " + maxAttempts + " tentativas falharam", lastException);
    }

    private long calculateBackoffDelay(int attempt) {
        // Backoff exponencial: delay * 2^(attempt-1), com limite máximo
        long delay = initialDelayMs * (long) Math.pow(2, attempt - 1);
        return Math.min(delay, maxDelayMs);
    }

    @FunctionalInterface
    public interface RetryableOperation<T> {
        T execute() throws Exception;
    }
}